File: scripts/reference-solutions/min-stack.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    q = int(data[idx])
    idx += 1
    stack = []
    mins = []
    out = []
    for _ in range(q):
        op = data[idx]
        idx += 1
        if op == b"push":
            x = int(data[idx])
            idx += 1
            stack.append(x)
            mins.append(x if not mins or x <= mins[-1] else mins[-1])
        elif op == b"pop":
            stack.pop()
            mins.pop()
        elif op == b"top":
            out.append(str(stack[-1]))
        else:  # getMin
            out.append(str(mins[-1]))
    sys.stdout.write("\n".join(out))
    if out:
        sys.stdout.write("\n")


main()

File: scripts/reference-solutions/min-stack.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int q;
  cin >> q;

  vector<long long> stack;
  vector<long long> mins;
  string op;
  string out;
  for (int i = 0; i < q; ++i) {
    cin >> op;
    if (op == "push") {
      long long x;
      cin >> x;
      stack.push_back(x);
      mins.push_back(mins.empty() || x <= mins.back() ? x : mins.back());
    } else if (op == "pop") {
      stack.pop_back();
      mins.pop_back();
    } else if (op == "top") {
      out += to_string(stack.back());
      out += '\n';
    } else {  // getMin
      out += to_string(mins.back());
      out += '\n';
    }
  }

  cout << out;
  return 0;
}
